File: dotnet/Examples/ControlClientUpdatingJSONTopics.cs
/**
 * Copyright © 2016 Push Technology Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using PushTechnology.ClientInterface.Client.Factories;
using PushTechnology.ClientInterface.Client.Features.Control.Topics;
using PushTechnology.ClientInterface.Client.Session;
using PushTechnology.ClientInterface.Client.Topics;
using PushTechnology.ClientInterface.Data.JSON;
using PushTechnology.ClientInterface.IO.CBOR;

namespace Examples {

    /// <summary>
    /// This example shows a control client creating a JSON topic and sending updates to it.
    ///
    /// There will be a topic for each currency for which rates are provided. The topic will be created under the FX
    /// topic - so, for example FX/GBP will contain a map of all rate conversions from the base GBP currency. The rates
    /// are represented as string decimal values (e.g. "12.457").
    ///
    /// The <c>addRates</c> method shows how to create a new rates topic, specifying its initial map of values.
    /// The <c>changeRates</c> method which takes a map shows how to completely replace the set of rates for a currency
    /// with a new map of rates.
    ///
    /// The <c>changeRates</c> method which takes a string shows an alternative mechanism where the new rates are
    /// simply supplied as a JSON string.
    ///
    /// Either of the changeRates methods could be used and after the first usage for any topic the values is cached,
    /// and so subsequent set calls can compare with the last value and send only the differences to the server.
    /// </summary>
    public class ControlClientUpdatingJSONTopics {
        private static readonly string rootTopic = "FX";
        private readonly ISession session;
        private volatile IValueUpdater<IJSON> valueUpdater;
        private readonly ITopicControl topicControl;
        private readonly IJSONDataType jsonDataType = Diffusion.DataTypes.JSON;
        private readonly CBORReader reader;

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="serverUrl">for example "ws://diffusion.example.com:80"</param>
        public ControlClientUpdatingJSONTopics( string serverUrl ) {
            session = Diffusion.Sessions.Principal( "control" ).Password( "password" ).Open( serverUrl );

            topicControl = session.GetTopicControlFeature();

            // Register as an updater for all topics under the root
            session.GetTopicUpdateControlFeature().RegisterUpdateSource( rootTopic,
                new MyTopicUpdateSource( valueUpdater ) );
        }

        /// <summary>
        /// Add a new rates topic.
        /// </summary>
        /// <param name="currency">the base currency</param>
        /// <param name="values">the full map of initial rates values</param>
        /// <param name="callback"> reports outcome</param>
        public void AddRates( string currency, IDictionary<string, string> values,
            ITopicControlAddContextCallback<string> callback ) {
            topicControl.AddTopic( RateTopicName( currency ), TopicType.JSON, MapToJSON( values ), currency, callback );
        }

        /// <summary>
        /// Update and existing rates topic, replacing the rates mappings with a new set of mappings.
        /// </summary>
        /// <param name="currency">the base currency</param>
        /// <param name="values">the mew rates values</param>
        /// <param name="callback">reports outcome</param>
        public void ChangeRates( string currency, IDictionary<string, string> values,
            ITopicUpdaterUpdateContextCallback<string> callback ) {
            if ( valueUpdater == null ) {
                throw new InvalidOperationException( "Not registered as updater" );
            }

            valueUpdater.Update( RateTopicName( currency ), MapToJSON( values ), currency, callback );
        }

        /// <summary>
        /// Update an existing rates topic, replacing the rates mappings with a new set of mappings specified as a JSON
        /// string, for example {"USD":"123.45","HKD":"456.3"}.
        /// </summary>
        /// <param name="currency">the base currency</param>
        /// <param name="jsonString">a JSON string specifying the map of currency rates</param>
        /// <param name="callback">reports outcome</param>
        public void ChangeRates( string currency, string jsonString,
            ITopicUpdaterUpdateContextCallback<string> callback ) {
            if ( valueUpdater == null ) {
                throw new InvalidOperationException( "Not registered as updater" );
            }

            valueUpdater.Update( RateTopicName( currency ), jsonDataType.FromJSONString( jsonString ), currency,
                callback );
        }

        /// <summary>
        /// Convert a given map to a JSON object.
        /// </summary>
        /// <param name="values">the map of values to be converted</param>
        /// <returns></returns>
        private IJSON MapToJSON( IDictionary<string, string> values ) {
            using ( var stream = new MemoryStream() ) {

                var writer = new CBORWriter( stream );

                writer.WriteObject();
                for ( int i = 0; i < values.Count; i++ ) {
                    writer.Write( values.ElementAt( i ).Key );
                    writer.Write( values.ElementAt( i ).Value );
                }
                writer.WriteBreak();

                return jsonDataType.ReadValue( stream.ToArray() );
            }
        }


        /// <summary>
        /// Remove a rates entry (remove its topic) and clear cached value for the topic.
        /// </summary>
        /// <param name="currency">the currency to be removed</param>
        /// <param name="callback">reports the outcome</param>
        public void removeRates( string currency, ITopicControlRemoveContextCallback<string> callback ) {
            var topicName = RateTopicName( currency );
            if ( valueUpdater != null ) {
                valueUpdater.RemoveCachedValues( topicName );
            }

            topicControl.RemoveTopics( RateTopicName( currency ), currency, callback );
        }

        /// <summary>
        /// Close the session.
        /// </summary>
        public void close() {
            // Remove our topics and close session when done
            topicControl.RemoveTopics( rootTopic, new MyTopicControlRemoveCallback(session) );
        }

        /// <summary>
        /// Generate a hierarchical topic name for a rates topic. e.g. for currency=GBP would return "FX/GBP".
        /// </summary>
        /// <param name="currency"></param>
        /// <returns>the topic name</returns>
        private static string RateTopicName( string currency ) {
            if ( currency == null ) {
                throw new ArgumentNullException( currency );
            }
            return String.Format( "%s/%s", rootTopic, currency );
        }

        private class MyTopicUpdateSource : TopicUpdateSourceDefault {
            private IValueUpdater<IJSON> valueUpdater;

            public MyTopicUpdateSource( IValueUpdater<IJSON> valueUpdater ) {
                this.valueUpdater = valueUpdater;
            }

            public override void OnActive( string topicPath, ITopicUpdater updater ) {
                valueUpdater = updater.ValueUpdater<IJSON>();
            }
        }

        private class MyTopicControlRemoveCallback : ITopicControlRemoveCallback {
            private ISession session;

            public MyTopicControlRemoveCallback( ISession session ) {
                this.session = session;
            }

            public void OnDiscard() {
                session.Close();
            }

            public void OnTopicsRemoved() {
                session.Close();
            }
        }
    }
}

File: java/src/main/java/com/pushtechnology/diffusion/examples/ControlClientUpdatingJSONTopics.java
/*******************************************************************************
 * Copyright (C) 2016, 2019 Push Technology Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
package com.pushtechnology.diffusion.examples;

import static com.pushtechnology.diffusion.client.Diffusion.newTopicSpecification;
import static com.pushtechnology.diffusion.client.topics.details.TopicSpecification.REMOVAL;
import static com.pushtechnology.diffusion.client.topics.details.TopicType.JSON;
import static java.util.Objects.requireNonNull;
import static java.util.concurrent.TimeUnit.SECONDS;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeoutException;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.cbor.CBORFactory;
import com.fasterxml.jackson.dataformat.cbor.CBORGenerator;
import com.pushtechnology.diffusion.client.Diffusion;
import com.pushtechnology.diffusion.client.features.TopicUpdate;
import com.pushtechnology.diffusion.client.features.control.topics.TopicControl;
import com.pushtechnology.diffusion.client.session.Session;
import com.pushtechnology.diffusion.client.topics.details.TopicSpecification;
import com.pushtechnology.diffusion.client.topics.details.TopicType;
import com.pushtechnology.diffusion.datatype.json.JSON;
import com.pushtechnology.diffusion.datatype.json.JSONDataType;

/**
 * This example shows a control client creating a JSON topic and sending updates
 * to it.
 * <P>
 * There will be a topic for each currency for which rates are provided. The
 * topic will be created under the FX topic - so, for example FX/GBP will
 * contain a map of all rate conversions from the base GBP currency. The rates
 * are represented as string decimal values (e.g. "12.457").
 * <P>
 * The {@code addRates} method shows how to create a new rates topic, specifying
 * its initial map of values.
 * <P>
 * The {@code changeRates} method which takes a map shows how to completely
 * replace the set of rates for a currency with a new map of rates.
 * <P>
 * The {@code changeRates} method which takes a string shows an alternative
 * mechanism where the new rates are simply supplied as a JSON string.
 * <P>
 * Either of the changeRates methods could be used and after the first usage for
 * any topic the values is cached, and so subsequent set calls can compare with
 * the last value and send only the differences to the server.
 *
 * @author DiffusionData Limited
 * @since 5.7
 * @see ClientConsumingJSONTopics
 */
public final class ControlClientUpdatingJSONTopics {

    private static final String ROOT_TOPIC = "FX";

    private final Session session;
    private final TopicControl topicControl;
    private final CBORFactory cborFactory = new CBORFactory();
    private final JSONDataType jsonDataType = Diffusion.dataTypes().json();

    /**
     * Constructor.
     *
     * @param serverUrl for example "ws://diffusion.example.com:80"
     */
    public ControlClientUpdatingJSONTopics(String serverUrl)
        throws InterruptedException, ExecutionException, TimeoutException {

        cborFactory.setCodec(new ObjectMapper());

        session =
            Diffusion.sessions().principal("control").password("password")
                .open(serverUrl);

        topicControl = session.feature(TopicControl.class);

        // Create the root topic that will remove itself when the session closes
        final TopicSpecification specification =
            newTopicSpecification(TopicType.STRING)
                .withProperty(REMOVAL,
                "When this session closes remove '?" + ROOT_TOPIC + "//'");

        topicControl.addTopic(ROOT_TOPIC, specification).get(5, SECONDS);
    }

    /**
     * Add a new rates topic and sets initial values.
     *
     * @param currency the base currency
     * @param values the full map of initial rates values
     * @return a CompletableFuture that completes when a response is received
     *         from the server
     */
    public CompletableFuture<?> addRates(
        String currency,
        Map<String, String> values)
        throws InterruptedException, ExecutionException, TimeoutException,
        IOException {

        topicControl.addTopic(rateTopicName(currency), JSON).get(5, SECONDS);

        return changeRates(currency, values);
    }

    /**
     * Update an existing rates topic, replacing the rates mappings with a new
     * set of mappings.
     *
     * @param currency the base currency
     * @param values the new rates values
     * @return a CompletableFuture that completes when a response is received
     *         from the server
     * @throws IOException if unable to convert rates map
     */
    public CompletableFuture<?> changeRates(
        String currency,
        Map<String, String> values) throws IOException {

        return session.feature(TopicUpdate.class).set(
            rateTopicName(currency),
            JSON.class,
            mapToJSON(values));
    }

    /**
     * Update an existing rates topic, replacing the rates mappings with a new
     * set of mappings specified as a JSON string, for example
     * {"USD":"123.45","HKD":"456.3"}.
     *
     * @param currency the base currency
     * @param jsonString a JSON string specifying the map of currency rates
     * @return a CompletableFuture that completes when a response is received
     *         from the server
     * @throws IOException if unable to convert string
     */
    public CompletableFuture<?> changeRates(
        String currency,
        String jsonString)
        throws IllegalArgumentException, IOException {

        return session.feature(TopicUpdate.class).set(
            rateTopicName(currency),
            JSON.class,
            jsonDataType.fromJsonString(jsonString));
    }

    /**
     * Convert a given map to a JSON object.
     */
    private JSON mapToJSON(Map<String, String> values) throws IOException {
        // Use the third-party Jackson library to write out the values map as a
        // CBOR-format binary.
        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
        final CBORGenerator generator = cborFactory.createGenerator(baos);
        generator.writeObject(values);
        return jsonDataType.readValue(baos.toByteArray());
    }

    /**
     * Remove a rates entry (removes its topic).
     *
     * @param currency the currency
     *
     * @return a CompletableFuture that completes when a response is received
     *         from the server
     */
    public CompletableFuture<?> removeRates(String currency) {

        final String topicName = rateTopicName(currency);

        return topicControl.removeTopics(topicName);
    }

    /**
     * Close the session.
     */
    public void close() {
        session.close();
    }

    /**
     * Generate a hierarchical topic name for a rates topic.
     * <P>
     * e.g. for currency=GBP would return "FX/GBP".
     *
     * @param currency the currency
     * @return the topic name
     */
    private static String rateTopicName(String currency) {
        return String.format("%s/%s", ROOT_TOPIC, requireNonNull(currency));
    }

}
